src/halo_regions.jl

```diff
@@ -1,93 +1,70 @@
-    Nx, Ny, Nz = grid.Nx, grid.Ny, grid.Nz
-    Hx, Hy, Hz = grid.Hx, grid.Hy, grid.Hz
-
-    @views @inbounds @. f.parent[1:Hx,           :, :] = f.parent[Nx+1:Nx+Hx, :, :]
-    @views @inbounds @. f.parent[Nx+Hx+1:Nx+2Hx, :, :] = f.parent[1+Hx:2Hx,   :, :]
-
-    if bcs.y.left == PeriodicBC
-        @views @inbounds @. f.parent[:, 1:Hy,           :] = f.parent[:, Ny+1:Ny+Hy, :]
-        @views @inbounds @. f.parent[:, Ny+Hy+1:Ny+2Hy, :] = f.parent[:, 1+Hy:2Hy,   :]
-    elseif bcs.y.left == NoFluxBC
-        @views @inbounds @. f.parent[:, 1:Hy,           :] = f.parent[:, 1+Hy:2Hy,   :]
-        @views @inbounds @. f.parent[:, Ny+Hy+1:Ny+2Hy, :] = 0
-
-        f.parent[:, 1+Hy, :] .= 0  # Enforce v=0 at the wall
-    end
+"Fill north and soϕth halo regions."
+function fill_y_halo_regions!(ϕ, bcs, grid)
+    fill_south_halo!(ϕ, bcs.left, grid.Hy, grid.Ny)
+    fill_north_halo!(ϕ, bcs.right, grid.Hy, grid.Ny)
+    return nothing
 end
 
-function fill_halo_region!(grid::Grid, ::Val{:w}, bcs::FieldBoundaryConditions, f)
-    Nx, Ny, Nz = grid.Nx, grid.Ny, grid.Nz
-    Hx, Hy, Hz = grid.Hx, grid.Hy, grid.Hz
-
-    @views @inbounds @. f.parent[1:Hx,           :, :] = f.parent[Nx+1:Nx+Hx, :, :]
-    @views @inbounds @. f.parent[Nx+Hx+1:Nx+2Hx, :, :] = f.parent[1+Hx:2Hx,   :, :]
-
-    if bcs.y.left == PeriodicBC
-        @views @inbounds @. f.parent[:, 1:Hy,           :] = f.parent[:, Ny+1:Ny+Hy, :]
-        @views @inbounds @. f.parent[:, Ny+Hy+1:Ny+2Hy, :] = f.parent[:, 1+Hy:2Hy,   :]
-    elseif bcs.y.left == NoFluxBC
-        @views @inbounds @. f.parent[:, 1:Hy,           :] = f.parent[:, 1+Hy:2Hy,   :]
-        @views @inbounds @. f.parent[:, Ny+Hy+1:Ny+2Hy, :] = -f.parent[:, Ny+1:Ny+Hy, :]
-    end
+"Fill top and bottom halo regions."
+function fill_z_halo_regions!(ϕ, bcs, grid)
+    fill_top_halo!(ϕ, bcs.left, grid.Hz, grid.Nz)
+    fill_bottom_halo!(ϕ, bcs.right, grid.Hz, grid.Nz)
+    return nothing
 end
 
-function fill_halo_region!(grid::Grid, ::Val{:T}, bcs::FieldBoundaryConditions, f)
-    Nx, Ny, Nz = grid.Nx, grid.Ny, grid.Nz
-    Hx, Hy, Hz = grid.Hx, grid.Hy, grid.Hz
-
-    @views @inbounds @. f.parent[1:Hx,           :, :] = f.parent[Nx+1:Nx+Hx, :, :]
-    @views @inbounds @. f.parent[Nx+Hx+1:Nx+2Hx, :, :] = f.parent[1+Hx:2Hx,   :, :]
-
-    if bcs.y.left == PeriodicBC
-        @views @inbounds @. f.parent[:, 1:Hy,           :] = f.parent[:, Ny+1:Ny+Hy, :]
-        @views @inbounds @. f.parent[:, Ny+Hy+1:Ny+2Hy, :] = f.parent[:, 1+Hy:2Hy,   :]
-    elseif bcs.y.left == NoFluxBC
-        @views @inbounds @. f.parent[:, 1:Hy,           :] = f.parent[:, 1+Hy:2Hy,   :]
-        @views @inbounds @. f.parent[:, Ny+Hy+1:Ny+2Hy, :] = f.parent[:, Ny+1:Ny+Hy, :]
-    end
+"Fill halo regions in x, y, and z for a given field."
+function fill_halo_regions!(field::AbstractArray, fieldbcs, grid)
+    fill_x_halo_regions!(field, fieldbcs.x, grid)
+    fill_y_halo_regions!(field, fieldbcs.y, grid)
+    #fill_z_halo_regions!(field, fieldbcs.z, grid)
```

Comment:
@ali-ramadhan, uncommenting this line leads to test failure. Do you mind taking a look? I couldn't figure out why immediately.

Reply:
Will do.

Reply:
Hmmm, yeah I'm weirded out now because I tried commenting it out like you did then filled in the halos manually but it didn't change any of `u, v, w, T, S`...

I'll try a few more things.

Reply:
Hmm.... 😳

Reply:
Shouldn't we get a bounds error when `Hz=0`?

Reply:
With `Nx, Ny, Nz = 32, 32, 32` and `Hx, Hy, Hz=1, 1, 0` everything seems well-defined
```julia
ϕ.parent[:, :, 1:Hz] -> 34×34×0 Array{Float64,3} (no array basically)
ϕ.parent[:, :,  1+Hz:1+Hz] -> 34×34×1 Array{Float64,3}
ϕ.parent[:, :, Nz+Hz+1:Nz+2Hz] -> 34×34×0 Array{Float64,3}
ϕ.parent[:, :, Nz+Hz:Nz+Hz] -> 34×34×1 Array{Float64,3}

ϕ.parent[:, :, 1:Hz+1] -> 34×34×1 Array{Float64,3}
ϕ.parent[:, :, Nz+Hz+1:Nz+2Hz] -> 34×34×0 Array{Float64,3}
```
and `ϕ.parent[:, :, Nz+Hz+100:Nz+Hz+100]` returns a `BoundsError` so that's assuring.

Reply:
Interesting: it seems to be setting `u[:, :, 1] = 0, v[:, :, 1] = 0, and w[:, :, 2] = 0`...

Reply:
Changing it to 
```julia
 fill_west_halo!(ϕ, ::NPBC, H, N) = @views @. ϕ.parent[1:H, :, :] = 0
fill_south_halo!(ϕ, ::NPBC, H, N) = @views @. ϕ.parent[:, 1:H, :] = 0
  fill_top_halo!(ϕ, ::NPBC, H, N) = @views @. ϕ.parent[:, :, 1:H] = 0
```
fixed it for me. Just running tests before committing.

Makes sense to me as halos are not responsible for e.g. enforcing `w=0` at the top, the operators already do that.

What do you think?

Reply:
No, this does not make sense to me. 

The code you have pasted relates to the `NoPenetration` boundary condition. This boundary condition is used for velocities fields that are normal to a boundary. For example, in the horizontally-periodic models this is the z-boundary condition for w. In the channel this is also the y boundary condition for `v`.

It seems to me that something is wrong if `w[1, :, :]` is not equal to zero. This is the value of w at the upper surface, which should be zero.

Please do not commit this code, as it does not make sense to me.